productivity-suite/piercing-library/src/reframed-client.js

```diff
@@ -0,0 +1,241 @@
+  get: () => {
+    return reframedContainer;
+  },
+});
+
+//const originalDocumentHead = document.head;
+Object.defineProperty(document, "head", {
+  get: () => {
+    return reframedContainer;
+  },
+});
+
+//const originalDocumentStylesheets = document.stylesheets;
+Object.defineProperty(document, "stylesheets", {
+  get: () => {
+    return reframedDocument.stylesheets;
+  },
+});
+
+// window.location is read-only and non-configurable, so we can't patch it
+//
+// additionally in a browsing context with one or more iframes, the history
+// all frames contribute to the joint history: https://www.w3.org/TR/2011/WD-html5-20110525/history.html#joint-session-history
+// this means that we need to be careful not to add duplicate entries to the
+// history stack via pushState within the iframe as that would double the
+// number of history entries that back/forward button would have to work through
+//
+// therefore we do the following:
+// - intercept all history.pushState history.replaceState calls and replay
+//   them in the main window
+// - update the window.location within the iframe via history.replaceState
+// - intercept window.addEventListener('popstate', ...) registration and forward it onto the main window
+const originalHistoryFns = new Map();
+["back", "forward", "go", "pushState", "replaceState"].forEach((prop) => {
+  originalHistoryFns.set(prop, window.history[prop]);
+  Object.defineProperty(window.history, prop, {
+    get: () => {
+      return function reframedHistoryGetter() {
+        console.log(
+          prop,
+          "history length",
+          reframedWindow.history.length,
+          window.history.length
+        );
+
+        switch (prop) {
+          case "pushState": {
+            Reflect.apply(
+              originalHistoryFns.get("replaceState"),
+              window.history,
+              arguments
+            );
+            reframedWindow.history.pushState(...arguments);
+            break;
+          }
+          case "replaceState": {
+            Reflect.apply(
+              originalHistoryFns.get("replaceState"),
+              window.history,
+              arguments
+            );
+            reframedWindow.history.replaceState(...arguments);
+            break;
+          }
+          default: {
+            Reflect.apply(
+              reframedWindow.history[prop],
+              reframedWindow.history,
+              arguments
+            );
+          }
+        }
+      };
+    },
+  });
+});
+
+// keep window.location and history.state in sync with the ones in the parent window
+reframedWindow.addEventListener("popstate", () => {
+  Reflect.apply(originalHistoryFns.get("replaceState"), window.history, [
+    reframedWindow.history.state,
+    null,
+    reframedWindow.location.href,
+  ]);
+});
+
+["length", "scrollRestoration", "state"].forEach((prop) => {
+  Object.defineProperty(window.history, prop, {
+    get: () => {
+      return Reflect.get(reframedWindow.history, prop);
```

Comment:
is the `Reflect.get` necessary here? wouldn't `reframedWindow.history[prop]` work just as well?

Reply:
Not sure, @IgorMinar does `Reflect` give us anything special here?

Reply:
I don't think so...

Reply:
I tried to use Reflect everywhere where I was accessing the foreign context. This was done for improving code readability. Metaprogramming and cross context code can be hard to read that's why I used Reflect api as an aid.

Reply:
I don't recall any part of this code actually requiring Reflect apis. But if we stick with it we should use it consistently.

Reply:
One more reason for Reflect was that it's also consistent with the use of Object.defineProperty pattern in this file.